### ISSUE TITLE
Flaky: policy is never assigned to agent

### ISSUE THREAD

**Issue by mtojek** [elastic/beats#28210]
Hi Team,

it happens from time to time and I suppose we have already fixed this: https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/master/913/pipeline/

Logs and metrics are in the "Artifacts" tab.

If you need internal filebeat and metricbeat logs of Elastic Agent, let me know and I will tell you how to access them (stored in beats-ci-temp bucket).

It affects master branch, spotted for 7.14.0-SNAPSHOT

**Comment by mtojek** [comment]
Most recent bug: https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/master/969/pipeline

**Comment by lykkin** [comment]
@mtojek thanks, i'll take a crack at this tomorrow to see if i can get to the bottom of it

**Comment by lykkin** [comment]
from what i can tell it accurately updates the policy through kibana, then fleet server seems to ignore the update. what @ruflin said sounds plausible, as it might stall out the fleet server from picking up the policy change. adding more logs sounds like a great idea, and i'll rope @ruflin in when i circle back around to trying to repro this.

**Comment by mtojek** [comment]
Hi @lykkin and the Team,

the issue is striking back again: https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations/detail/master/1317/pipeline

Let me know if you need more reference data.

**Comment by ruflin** [comment]
Based on the Jenkins logs this seems to be 7.16.0-SNAPSHOT. I was hoping a few fixes we did for 7.16 which are related to heavy load on ES or fleet-server overloading could fix this. There is one fix that only landed in 7.16.1 and based on the container that is started not such which version was tested. At the same time, the error we are seeing is not a connection error to Elasticsearch. 

With 7.16 quite a few improvements went in around logging and we now have also diagnostics in place. @lykkin Can you dig into the logs if you see something obvious? Ideally we would even run diagnostics on the Elastic Agent that struggles to see the exact state but this will be trickier as it is inside the container.

@kpollich You have recently looked into issues around duplicated fleet-server inputs so I wonder if this could be in any related maybe? Have nothing found to proof this but was wondering if this might ring a bell on your end.

@jlind23 We should get the ground of this one.

**Comment by kpollich** [comment]
> @kpollich You have recently looked into issues around duplicated fleet-server inputs so I wonder if this could be in any related maybe? Have nothing found to proof this but was wondering if this might ring a bell on your end.

Yes this part in particular sounds very much like the duplicate Fleet Server input issue:

> from what i can tell it accurately updates the policy through kibana, then fleet server seems to ignore the update

When we upgraded Fleet Server policies from older versions (prior to the Fleet Server package having a `policy_template` defined) we'd incur the duplicated input. Would be fixed in 7.16.1 by https://github.com/elastic/kibana/pull/119925.

**Comment by mtojek** [comment]
Hi folks, thanks for investigating the problem. Just wanted to remind you that I spotted the issue today, so I assume the CI used latest Docker images.

**Comment by lykkin** [comment]
heya, sorry about the lapse in comms, picking this back up for some more investigation. will update when i find something interesting.

**Comment by lykkin** [comment]
digging into the logs doesn't have anything conclusive. the [fleet-server logs](https://beats-ci.elastic.co/job/Ingest-manager/job/integrations/job/master/1357/artifact/build/elastic-stack-dump/network_traffic/logs/fleet-server.log) are pretty sparse, and don't track any changes around the time the reassign request happens.

tracing [the script logs](https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/branches/master/runs/1357/nodes/592/steps/8986/log/?start=0), we see the reassign request being made around 15:29:43
```
[2021-12-20T15:39:43.337Z] 2021/12/20 15:39:43 DEBUG reassigning original policy back to agent...
[2021-12-20T15:39:43.337Z] 2021/12/20 15:39:43 DEBUG PUT http://127.0.0.1:5601/api/fleet/agents/d8e9da95-a991-4daa-9231-94a4f11507a6/reassign
[2021-12-20T15:39:45.246Z] 2021/12/20 15:39:45 DEBUG GET http://127.0.0.1:5601/api/fleet/agents/d8e9da95-a991-4daa-9231-94a4f11507a6
```

which the [kibana logs](https://beats-ci.elastic.co/job/Ingest-manager/job/integrations/job/master/1357/artifact/build/elastic-stack-dump/network_traffic/logs/kibana.log) show were successful
```json
{"type":"response","@timestamp":"2021-12-20T15:39:43+00:00","tags":["access:fleet-all"],"pid":1223,"method":"put","statusCode":200,"req":{"url":"/api/fleet/agents/d8e9da95-a991-4daa-9231-94a4f11507a6/reassign","method":"put","headers":{"host":"127.0.0.1:5601","user-agent":"Go-http-client/1.1","content-length":"55","content-type":"application/json","kbn-xsrf":"7.16.0","accept-encoding":"gzip"},"remoteAddress":"172.18.0.1","userAgent":"Go-http-client/1.1"},"res":{"statusCode":200,"responseTime":1871,"contentLength":2},"message":"PUT /api/fleet/agents/d8e9da95-a991-4daa-9231-94a4f11507a6/reassign 200 1871ms - 2.0B"}
{"type":"response","@timestamp":"2021-12-20T15:39:45+00:00","tags":["access:fleet-read"],"pid":1223,"method":"get","statusCode":200,"req":{"url":"/api/fleet/agents/d8e9da95-a991-4daa-9231-94a4f11507a6","method":"get","headers":{"host":"127.0.0.1:5601","user-agent":"Go-http-client/1.1","content-type":"application/json","kbn-xsrf":"7.16.0","accept-encoding":"gzip"},"remoteAddress":"172.18.0.1","userAgent":"Go-http-client/1.1"},"res":{"statusCode":200,"responseTime":9,"contentLength":1282},"message":"GET /api/fleet/agents/d8e9da95-a991-4daa-9231-94a4f11507a6 200 9ms - 1.3KB"}
```

it appears that [the corresponding agent update is missing](https://beats-ci.elastic.co/job/Ingest-manager/job/integrations/job/master/1357/artifact/build/elastic-stack-dump/network_traffic/logs/elastic-agent.log)

there's nothing suggestive in the ES or fleet-server logs that would explain the drop in communication from kibana to the agent. it might be that running fleet-server at higher verbosity logging might catch something next time this issue crops up.

**Comment by blakerouse** [comment]
@lykkin Can you check the `.fleet-policies` index to see if the policy was placed into the index? You should see the policy with the input twice in the index.

First one should be with `revision_idx: ${revision_num}` and `coordinator_idx: 0` (that one was inserted by Kibana), then you should see a follow up inserted by the Fleet Server Coordinator with `revision_idx: ${revision_num}` and `coordinator_idx: 1`.

**Comment by lykkin** [comment]
@blakerouse good idea! i was under the impression the ES instance these tests ran against was ephemeral. is there a way to get at the ES container it used?

unfortunately, i have had no luck reproducing this locally.

**Comment by jlind23** [comment]
Any update there?
